EIPS/eip-3332.md

```diff
@@ -0,0 +1,95 @@
+
+```
+// Assume that block.medgasprice is bound to MEDGASPRICE (0x46)
+
+function submitFraudProof(bytes calldata proof) public {
+  require(tx.gasprice <= maxGasPrice());
+  // process the fraud proof and provide a reward (if valid)
+}
+
+function maxGasPrice() public view returns (uint) {
+  return 3 * block.medgasprice;
+}
+```
+
+Given the contract implementation above a client would simply call `maxGasPrice` to determine the gas price to use when submitting a fraud proof. This particular implementation allows up to 3x the median gas price of the last block to be used.
+
+#### Naming note
+
+The name `MEDGASPRICE` was chosen because the median gas price of the network can only be calculated from the latest complete block. Thus transactions being executed should expect the median gas price to be calculated from the previous block.
+
+## Backwards Compatibility
+
+This proposal introduces a single new opcode. No backwards incompatibilities should be introduced.
+
+## Forwards Compatibility
+
+[EIP-1559](https://eips.ethereum.org/EIPS/eip-1559) plans to change the fee market in a number of ways. Most notably is the creation of a base fee that is burned. In this context an "inclusion fee" still exists as a part of the total fee. Consider the following two cases:
+
+##### Block sizes are increasing (all available gas is being consumed)
+
+In this case there will be bidding contention in the inclusion fee to incentivize miners to include transactions. A median gas price operator would still be helpful as an attacker could supply a high inclusion fee to bump honest transactions.
+
+##### Block sizes are decreasing (excess gas is available)
+
+In this case an attacker could specify a high inclusion fee to incentivize miners to include their transaction early in the block. Miners are incentivized to do so as including expensive transactions first reduces the risk of a revert (and partial refund) occurring.
+
+Given these two cases this EIP seems relevant in the context of EIP-1559.
+
+Post EIP-1559 `MEDGASPRICE (0x46)` should return the median `effective_gas_price` of the previous block.
+
+[EIP-3198](https://eips.ethereum.org/EIPS/eip-3198) is required for the above strategies to be implemented. With the inclusion of `BASEFEE (0x48)` a contract can subtract the `base_fee_per_gas` from the `effective_gas_price` to determine the inclusion fee per gas being paid for the transaction and thus implement an upper bound.
+
+## Security Considerations
+
+The strategy described for preventing front-running by setting an upper bound on the gas price of transactions has a few caveats:
+
+1. It relies on miners being impartial. Reordering transactions with the same gas price is a trivial means of defeating this strategy.
+2. It relies on the implicit behavior of miners ordering transactions of the same value in the order received. This is not standard behavior but seems to be general practice.
```

Comment:
This isn't true.  Miners "often" (from a security standpoint) sort transactions against gas price.  There are also established mechanisms for paying miners out of band.